oauth2_http/java/com/google/auth/oauth2/GdchCredentials.java

```diff
@@ -0,0 +1,530 @@
+      this.privateKeyId = privateKeyId;
+      return this;
+    }
+
+    public Builder setPrivateKey(PrivateKey privateKey) {
+      this.privateKey = privateKey;
+      return this;
+    }
+
+    public Builder setServiceIdentityName(String name) {
+      this.serviceIdentityName = name;
+      return this;
+    }
+
+    public Builder setTokenServerUri(URI tokenServerUri) {
+      this.tokenServerUri = tokenServerUri;
+      return this;
+    }
+
+    public Builder setHttpTransportFactory(HttpTransportFactory transportFactory) {
+      this.transportFactory = transportFactory;
+      return this;
+    }
+
+    public Builder setCaCertPath(String caCertPath) {
+      this.caCertPath = caCertPath;
+      return this;
+    }
+
+    public Builder setGdchAudience(URI apiAudience) {
+      this.apiAudience = apiAudience;
+      return this;
+    }
+
+    public String getProjectId() {
+      return projectId;
+    }
+
+    public String getPrivateKeyId() {
+      return privateKeyId;
+    }
+
+    public PrivateKey getPrivateKey() {
+      return privateKey;
+    }
+
+    public String getServiceIdentityName() {
+      return serviceIdentityName;
+    }
+
+    public URI getTokenServerUri() {
+      return tokenServerUri;
+    }
+
+    public HttpTransportFactory getHttpTransportFactory() {
+      return transportFactory;
+    }
+
+    public String getCaCertPath() {
+      return caCertPath;
+    }
+
+    public int getLifetime() {
+      return lifetime;
+    }
+
+    public GdchCredentials build() {
+      return new GdchCredentials(this);
+    }
+  }
+
+  private static String validateField(String field, String fieldName) throws IOException {
+    if (field == null || field.isEmpty()) {
+      throw new IOException(
+          String.format(
+              "Error reading GDCH service account credential from JSON, %s is misconfigured.",
+              fieldName));
+    }
+    return field;
+  }
+
+  /*
+   * Internal HttpTransportFactory for GDCH credentials.
+   *
+   * <p> GDCH authentication server could use a self-signed certificate, thus the client could
+   * provide the CA certificate path through the `ca_cert_path` in GDCH JSON file.
+   *
+   * <p> The TransportFactoryForGdch subclass would read the certificate and create a trust store,
+   * then use the trust store to create a transport.
+   *
+   * <p> If the GDCH authentication server uses well known CA certificate, then a regular transport
+   * would be set.
+   */
+  static class TransportFactoryForGdch implements HttpTransportFactory {
+    HttpTransport transport;
+
+    public TransportFactoryForGdch(String caCertPath) throws IOException {
+      setTransport(caCertPath);
+    }
+
+    @Override
+    public HttpTransport create() {
+      return transport;
+    }
+
+    private void setTransport(String caCertPath) throws IOException {
+      if (caCertPath == null || caCertPath.isEmpty()) {
+        this.transport = new NetHttpTransport();
```

Comment:
no it can never be a NetTransport. The location of the cert could be a default one.

Reply:
reverted the ca cert path to required

Reply:
optional is the way to go